src/marqo/s2_inference/hf_utils.py

```diff
@@ -16,15 +24,95 @@ def __init__(self, *args, **kwargs) -> None:
 
         if self.max_seq_length is None:
             self.max_seq_length = 128
+        self.model_properties = kwargs.get("model_properties", dict())
+        self.model_name = self.model_properties.get("name", None)
+
+        model_auth = kwargs.get(InferenceParams.model_auth, None)
+        if model_auth is not None:
+            self.model_auth = model_auth
+        else:
+            self.model_auth = None
 
     def load(self) -> None:
 
-        self.model = AutoModelForSentenceEmbedding(self.model_name).to(self.device)
-        self.tokenizer = AutoTokenizer.from_pretrained(self.model_name)
+        model_location_presence = ModelProperties.model_location in self.model_properties
+        path = self.model_properties.get("localpath", None) or self.model_properties.get("url", None)
+        # HF models can be loaded from 3 entries: path (url or localpath), model_name, or model_location
+        if (path is not None) + (self.model_name is not None) + (model_location_presence is True) != 1:
+            raise InvalidModelPropertiesError("Exactly one of `url`, `localpath` or `model_location`, `name` can be specified"
+                                              " in `model_properties` for `hf` models as they conflict with each other in model loading."
+                                              " Please ensure that exactly one of these is specified in `model_properties` and retry.")
+        elif path is not None:
+            if validators.url(path) is True:
+                self.model_path = download_model(url = path, download_dir=ModelCache.hf_cache_path)
+            elif os.path.isdir(path) or os.path.isfile(path):
+                self.model_path = path
+        elif self.model_name is not None:
+            # Loading from structured huggingface repo directly, token is required directly
+            self.model_path = self.model_name
+            self.model = AutoModelForSentenceEmbedding(self.model_path).to(self.device)
+            self.tokenizer = AutoTokenizer.from_pretrained(self.model_path)
+            return
+        elif model_location_presence is not None:
+            # This is a special case for huggingface models, where we can load a model directory from a repo
+            if ("hf" in self.model_properties["model_location"]) and ("repo_id" in self.model_properties["model_location"]["hf"]) and \
+                ("filename" not in self.model_properties["model_location"]["hf"]):
+                return self._load_from_private_hf_repo()
+            else:
+                self.model_path = self._download_from_repo()
 
+        else:
```

Comment:
I don't think we should have this logic duplicated twice. If line 41 logic is correct and we have an if statement per condition, then this `else` won't be hit. More code makes code more complex and fragile.